url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/patterns/interpreter/parsergenerator/parsertables/AbstractParserTables.java

```diff
@@ -0,0 +1,402 @@
+				if (s.length() > CELLWIDTH && it.hasNext())	// not last
+					CELLWIDTH = s.length() + 1;
+			}
+		}
+		
+		// print table header
+		dumpCell(" | ", CELLWIDTH, false, out);			
+		for (Iterator it = head.iterator(); it.hasNext(); )	{
+			String s = (String) it.next();
+			if (Token.isEpsilon(s))
+				s = "<EOF>";
+			dumpCell(s, CELLWIDTH, ! it.hasNext(), out);			
+		}
+		
+		out.println();
+		out.println(dummy(CELLWIDTH * (head.size() + 1), "_"));
+		
+		int i = 0;
+		for (Iterator it = table.iterator(); it.hasNext(); i++)	{
+			Map h = (Map) it.next();
+
+			dumpCell(Integer.toString(i)+" | ", CELLWIDTH, false, out);			
+			for (Iterator it2 = head.iterator(); it2.hasNext(); )	{
+				String symbol = (String) it2.next();
+
+				Integer intg = h != null ? (Integer) h.get(symbol) : null;
+				String digit = intg == null ? "-" : intg.equals(ParserTables.SHIFT) ? "SH" : intg.equals(ParserTables.ACCEPT) ? "AC" : intg.toString();
+
+				dumpCell(digit, CELLWIDTH, ! it2.hasNext(), out);
+			}
+			out.println();
+		}
+		
+		out.println();
+	}
+	
+	private void dumpCell(String digit, int width, boolean isLast, PrintStream out)	{
+		StringBuffer tab = new StringBuffer(" ");
+		for (int i = 0; i < width - 1 - digit.length(); i++)
+			tab.append(' ');
+			
+		String s = tab+digit;
+		if (!isLast && s.length() > width && width > 2)
+			s = s.substring(0, width);
+			
+		out.print(s);
+	}
+
+	private String dummy(int width, String ch)	{
+		StringBuffer sb = new StringBuffer();
+		for (int i = 0; i < width; i++)
+			sb.append(ch);
+		return sb.toString();
+	}
+
+
+	/** Reports count of rules, ternminals and nonterminals. */
+	public void report(PrintStream out)	{
+		out.println("rules: "+syntax.size());
+		out.println("terminals: "+terminals.size());
+		out.println("nonterminals: "+nonterminals.size());
+	}
+
+
+
+	
+	// Java source generation methods
+
+	
+	private transient Writer f = null;
+
+	private void fwrite(String line)
+		throws IOException
+	{
+		f.write(line, 0, line.length());
+	}
+	
+	/**
+		Generates a compilable source file "parserClassName.java".
+		If the class name contains package path, the file will be generated
+		within the corresponding directory, else within working directory.
+		@param parserClassName name of class to generate:
+				"Java" will generate "Java.java" in current directory,
+				"my.pkg.PT" will generate "my/pkg/PT.java" in current directory.
+		@return fileName if File could be written, else null.
+	*/
+	public String toSourceFile(String parserClassName)
+		throws IOException
+	{
+		if (parserClassName.endsWith("ParserTables") == false)
+			parserClassName = parserClassName+"ParserTables";
+			
+		String fileName = parserClassName.replace('.', File.separatorChar)+".java";
+		System.err.println("Writing Java Source to "+fileName);
+
+		File file = new File(fileName);
+		String path = file.getParent();
+		if (path != null && new File(path).exists() == false)
+			new File(path).mkdirs();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Confirm that `mkdirs` was successful. Handle any failures that happened. If you fail silently, then you might encounter errors that are harder to debug. See the following diffs on GitHub that make this code change: [diff1](https://github.com/qunarcorp/qmq/commit/12f1d9d6a57a02bc2415cb2af9f4388a77b12c02#diff-6b67be6e17f24fdeddb7c725ac8ec5600d6e63be0bdca0963a83d530e6d55e35L36), [diff2](https://github.com/apache/jmeter/commit/1ec61206779b3f84df18516682b4791bf015178f#diff-cd74304763c097fbc085ac2bc952b46ab3843ac825cacacad89e8064296e955aL330)

Learn more: [File.mkdirs in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/io/File.html#mkdirs--)